azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -937,6 +942,14 @@ def _run_op(self, op):
                         self.name, op.name, self.state
                     )
                 )
+                self.state = ReconnectState.LOGICALLY_CONNECTED
+                # We don't send this op down.  Instead, we send a new connect op down.  This way,
```

Comment:
What if we already were Logically Connected? Should we perhaps have an additional conditional block to prevent us from trying to connect when we already are (logically) connected?

Reply:
I like this train of thought, but it's taking me in a different direction.  For now, ConnectionLockStage is above this one, so we only get a ConnectOperation here if we're physically disconnected.  
In the longer term, your comment somewhere else in here makes me wonder if we want to have different types of connect ops:
PhysicalConnectOp - for calling into transport
LogicalConnectOp - for setting LogicallyConnected
GuaranteedPhysicalConnectOp - physical connect op with retry

or something like that.  Myabe PhysicalConnectOp is guaranteed and NonGuaranteedConnectOp is the one that isn't guaranteed.  I'm not sure.  

Either way, that's rambing and not well thouht out.  The code in here is good for now.

---
In reply to: [439846074](https://github.com/Azure/azure-iot-sdk-python/pull/576#discussion_r439846074) [](ancestors = 439846074)